File: javascript/042/main.js
/*
 * ProjectEuler 042
 * JavaScript / main.js
 * copyright (c) 2015 Susisu
 *
 * Node.js v5.0.0
 * `node main.js`
 */

"use strict";

var fs = require("fs");

function main() {
    fs.readFile("./words.txt", { "encoding": "utf8" }, function (err, res) {
        if (err) {
            console.error(err);
            process.exit(1);
            return;
        }
        var words =
            res.split("\n")[0]
                .split(",")
                .map(word => word.slice(1, -1))
                .map(wordValue)
                .filter(isTriangle);
        console.log(words.length);
    })
}

var CHAR_A = "A".charCodeAt(0);

function wordValue(word) {
    var value = 0;
    for (var i = 0; i < word.length; i++) {
        value += word.charCodeAt(i) - CHAR_A + 1;
    }
    return value;
}

function isTriangle(n) {
    var t = n * 2;
    var r = Math.sqrt(t);
    for (var i = 1; i < t; i++) {
        if (t % i === 0 && t / i === i + 1) {
            return true;
        }
    }
    return false;
}

main();

File: ocaml/042/main.ml
(*
    ProjectEuler 042
    OCaml / main.ml
    copyright (c) 2016 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main -linkpkg -package core -thread main.ml`
*)

open Core

let alph_position c = Char.code c - Char.code 'A' + 1

let word_value w =
    let ps = List.map alph_position @@ Core_string.to_list w in
    List.fold_left (+) 0 ps

let is_triangle n =
    let m = n * 2 in
    let r = int_of_float @@ floor @@ sqrt @@ float_of_int m in
    r * (r + 1) = m

let _ =
    let ic = open_in "./words.txt" in
    try
        let raw_words = input_line ic in
        let words = List.map (fun x -> Core_string.slice x 1 (-1))
            @@ Core_string.split raw_words ',' in
        let values = List.map word_value words in
        let triangles = List.filter is_triangle values in
        print_int (List.length triangles);
        print_newline ()
    with
    | err ->
        close_in_noerr ic;
        raise err
